content/en/blog/_posts/2024-04-23-recursive-read-only-mounts.md

```diff
@@ -0,0 +1,97 @@
+---
+layout: blog
+title: 'Kubernetes 1.30: Read-only volume mounts can be finally literally read-only'
+date: 2024-04-23
+slug: recursive-read-only-mounts
+---
+
+**Author:** Akihiro Suda (NTT)
+
+Kubernetes 1.30 added alpha support for "recursive read-only mounts".
+
+Prior to this release, read-only volume mounts were surprisingly not really read-only under specific conditions.
+
+### Read-only volume mounts are not really read-only
+
+Volume mounts are widely misunderstood.
+
+The following manifest is expected to make everything under `/mnt` in the containers read-only:
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  volumes:
+    - name: mnt
+      hostPath:
+        path: /mnt
+  containers:
+    - volumeMounts:
+        - name: mnt
+          mountPath: /mnt
+          readOnly: true
+```
+
+However, it actually leaves submounts beneath `/mnt` (e.g., NFS or removable media) writable.
+
+### New mount option: recursiveReadOnly
+
+Kubernetes 1.30 added a new mount option `recursiveReadOnly` so as to make submounts recursively read-only.
+
+The option can be enabled as follows:
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  volumes:
+    - name: mnt
+      hostPath:
+        path: /mnt
+  containers:
+    - volumeMounts:
+        - name: mnt
+          mountPath: /mnt
+          readOnly: true
+          # NEW
+          # Possible values are `Enabled`, `IfPossible`, and `Disabled`.
+          # Needs to be specified in conjunction with `readOnly: true`.
+          recursiveReadOnly: Enabled
+```
+
+This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag
+using [`mount_setattr(2)`](https://man7.org/linux/man-pages/man2/mount_setattr.2.html).
+
+### Availability
+
+To enable `recursiveReadOnly` mounts, the following components have to be used:
+
+* Kubernetes: v1.30 or later, with the `RecursiveReadOnlyMounts`
+  [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) enabled.
+  As of v1.30, the gate is marked as alpha.
+
+* CRI runtime:
+  * containerd: v2.0 or later
+
+* OCI runtime:
+  * runc: v1.1 or later
+  * crun: v1.8.6 or later
+
+* Linux kernel: v5.12 or later
+
+### What's next?
+
+The feature gate will be promoted to beta and eventually GA in future releases of Kubernetes
+so that users no longer need to enable the gate.
```

Comment:
```suggestion
Kubernetes SIG Node hope - and expect - that the feature will be promoted to beta and eventually
general availability (GA) in future releases of Kubernetes, so that users no longer need to enable
the feature gate manually.
```

You need the “we're hoping” unless SIG Architecture have committed to the graduation to stable.

Reply:
done